tripal/tripal
# configurable mechanism for hiding fields when they have an "empty" value

# New Feature

### Closes #1970 

### Tripal Version: 4.x

## Description
This PR provides a configurable mechanism for hiding fields when they have an "empty" value.
For text fields this would generally be an empty string, for integer fields it might be zero or -1,
for boolean fields it might be FALSE.
Being configurable, however, we can specify some other value, for example, there is a CV term `no_rank` that can be used for the infraspecific type on an organism, and we might want to hide that.

## Testing?
Note - because the change in commit https://github.com/tripal/tripal/pull/1971/commits/44d3dce11e1e6120560e95e524976541cc63a774 could have unintended consequences, I recommend putting a `dpm()` in `tripal/src/Entity/TripalEntity.php` line 566 so we know when the change is having an effect, do this after step 5 below
```
          // Keep track of elements that have no value.
          foreach ($prop_values as $i => $prop_value) {
            $prop_value_value = $prop_value->getValue();
if (!$prop_value_value and !is_null($prop_value_value)) {
  dpm($prop_value_value, "This value would have been removed in 4.x for field $field_name delta $delta");
}
            if (is_null($prop_value_value)) {
              // A given delta should only be present once here.
              if (!array_key_exists($field_name, $delta_remove) or !in_array($delta, $delta_remove[$field_name])) {
                $delta_remove[$field_name][] = $delta;
              }
              continue;
            }
          }
```
1. Build docker on `4.x` branch
2. Import type collection "Expression Content Types"
3. Enter a publication (tests text, string, and boolean (Is Obsolete) fields). Only enter required fields, leave others blank.
![2024-09-13_pub1](https://github.com/user-attachments/assets/f939c284-e707-4d25-a5db-ada64abe99b8)
4. Enter an array design which has lots of integer fields. Enter various values including zero, and leave some blank.
![2024-09-13_ad1](https://github.com/user-attachments/assets/f9e11864-1b6e-4269-9e77-94ed87ec8e07)
5. Checkout this branch `git checkout tv4g1-issue1970-hide-field-configuration` and make the `dpm()` addition described above
6. Clear cache just to be safe `docker exec -it CONTAINERNAME /bin/bash -c "drush cache:rebuild"`
6. View the publication, it should appear the same as before by default.
7. Try changing some of the settings Tripal -> Page Structure -> Publication -> Manage display
for example 
- series name change to "Never Hide"
- change the "Is Obsolete" as shown here to not display if FALSE
![2024-09-13_pub3](https://github.com/user-attachments/assets/e944900d-b96d-4ae4-95f6-193e144c6ed9)
- Unique local identifier change to "Hide if empty or equal to a specific value" and for the value use what you entered.
![2024-09-13_pub4](https://github.com/user-attachments/assets/29092bd0-a3d6-4cc9-ab0c-10e9dceeaa20)
- Don't forget to hit save
8. Note that Series name now appears even though it has no value, and Unique local identifier is hidden.
![2024-09-13_pub2](https://github.com/user-attachments/assets/09ba81dd-5686-4ed8-8655-79a5f591f861)
9. For array design, change the settings for an integer field containing a zero to not be displayed
These are the test values I used
![1970_AD0](https://github.com/user-attachments/assets/5fd2a243-e2ee-43b6-966a-0bc9a67d1c14)
![2024-09-13_ad2](https://github.com/user-attachments/assets/6be9cb43-156c-4452-a1af-039a74d9370c)
10. Change the settings for an integer field containing a specific value to not be displayed
![2024-09-13_ad3](https://github.com/user-attachments/assets/22012948-099e-42c1-870b-d0417726c81c)
11. View the array design to confirm they are no longer displayed
![2024-09-13_ad4](https://github.com/user-attachments/assets/2c3efe11-49b8-4dcc-99ab-1a8cce4c5202)
12. Add a **Tripal** boolean field to some simple content type, for example contact.
13. Create a contact each with true and false
14. On 4.x the false value will not appear, because it is not stored in the drupal field table. On this branch it will be present.
15. Similarly test a tripal integer field and enter the value zero. On 4.x it will not be there, on this branch it will. This should be the only time that the dpm we added triggers
![1970_dpm](https://github.com/user-attachments/assets/d3398ec3-624f-431e-8f11-91a9e54c7f6c)
![1970_tripalfieldsfalse](https://github.com/user-attachments/assets/49450935-dfd9-42cc-b8e9-6c80ddfae5b3)
16. Unpublish and republish to make sure that works correctly.
😬 Null integers are published as zero. This is not related to this PR but is a new issue #1972
